src/java/org/apache/cassandra/db/ColumnFamilyStore.java

```diff
@@ -1276,60 +1372,97 @@ private static void setCommitLogUpperBound(AtomicReference<CommitLogPosition> co
             {
-                float usedOnHeap = Memtable.MEMORY_POOL.onHeap.usedRatio();
-                float usedOffHeap = Memtable.MEMORY_POOL.offHeap.usedRatio();
-                float flushingOnHeap = Memtable.MEMORY_POOL.onHeap.reclaimingRatio();
-                float flushingOffHeap = Memtable.MEMORY_POOL.offHeap.reclaimingRatio();
-                float thisOnHeap = largest.getAllocator().onHeap().ownershipRatio();
-                float thisOffHeap = largest.getAllocator().offHeap().ownershipRatio();
-                logger.debug("Flushing largest {} to free up room. Used total: {}, live: {}, flushing: {}, this: {}",
-                            largest.cfs, ratio(usedOnHeap, usedOffHeap), ratio(liveOnHeap, liveOffHeap),
-                            ratio(flushingOnHeap, flushingOffHeap), ratio(thisOnHeap, thisOffHeap));
-                largest.cfs.switchMemtableIfCurrent(largest);
+                largest = current;
+                largestRatio = ratio;
             }
+
+            liveOnHeap += onHeap;
+            liveOffHeap += offHeap;
+        }
+        CompletableFuture<Boolean> ret = new CompletableFuture<>();
+        double liveRatio = Math.max(liveOnHeap, liveOffHeap);
+
+        if (largest != null && liveRatio >= minOwnershipRatio)
+        {
+            float usedOnHeap = Memtable.MEMORY_POOL.onHeap.usedRatio();
+            float usedOffHeap = Memtable.MEMORY_POOL.offHeap.usedRatio();
+            float flushingOnHeap = Memtable.MEMORY_POOL.onHeap.reclaimingRatio();
+            float flushingOffHeap = Memtable.MEMORY_POOL.offHeap.reclaimingRatio();
+            float thisOnHeap = largest.getAllocator().onHeap().ownershipRatio();
+            float thisOffHeap = largest.getAllocator().offHeap().ownershipRatio();
+            logger.debug("Flushing largest {} to free up room, min ratio {}. Used total: {}, live: {}, flushing: {}, this: {}",
+                         largest.cfs, ratio(minOwnershipRatio), ratio(usedOnHeap, usedOffHeap), ratio(liveOnHeap, liveOffHeap),
+                         ratio(flushingOnHeap, flushingOffHeap), ratio(thisOnHeap, thisOffHeap));
+
+            ListenableFuture<CommitLogPosition> flushFut = largest.cfs.switchMemtableIfCurrent(largest);
+            flushFut.addListener(() -> {
+                try
+                {
+                    flushFut.get();
+                    ret.complete(true);
+                }
+                catch (Throwable t)
+                {
+                    ret.completeExceptionally(t);
+                }
+            }, MoreExecutors.directExecutor());
         }
+        else
+        {
+            if (largest == null)
+                logger.debug("Flushing of largest memtable, not done, no memtable found");
+            else if (liveRatio < minOwnershipRatio)
+                logger.debug("Flushing of largest memtable, not done, max live ratio {} less than min ratio {}", ratio(liveRatio), ratio(minOwnershipRatio));
```

Comment:
nit: I'm guessing the `ratio()` calculations are not expensive enough to warrant guarding this with `isDebugEnabled()`?